File: sdks/go/endpoint.go
package rrd

import (
	"errors"
	"fmt"
	"net"
	"net/url"
	"strings"
)

func loopbackURL(value string) (*url.URL, error) {
	parsed, err := url.Parse(value)
	if err != nil {
		return nil, fmt.Errorf("parse RRD URL: %w", err)
	}
	host := parsed.Hostname()
	loopback := host == "localhost"
	if address := net.ParseIP(host); address != nil {
		loopback = address.IsLoopback()
	}
	if parsed.Scheme != "http" || !loopback || parsed.User != nil || parsed.RawQuery != "" ||
		parsed.Fragment != "" {
		return nil, errors.New(
			"RRD Go client permits only credential-free loopback HTTP before TLS qualification",
		)
	}
	if !strings.HasSuffix(parsed.Path, "/") {
		parsed.Path += "/"
	}
	return parsed, nil
}

File: sdks/typescript/src/endpoint.ts
/** Current credential-free loopback endpoint profile. */

import { RrdClientError } from "./error.js";

export interface ClientConfig {
  baseUrl: string;
  instance: string;
  requestTimeoutMs?: number;
  maxAttempts?: number;
  maxResponseBytes?: number;
  fetch?: typeof globalThis.fetch;
}

export function loopbackUrl(value: string): URL {
  const url = new URL(value);
  if (
    url.protocol !== "http:" ||
    !["localhost", "127.0.0.1", "[::1]"].includes(url.hostname) ||
    url.username ||
    url.password ||
    url.search ||
    url.hash
  ) {
    throw new RrdClientError(
      "RRD TypeScript client permits only credential-free loopback HTTP before TLS qualification",
    );
  }
  if (!url.pathname.endsWith("/")) url.pathname += "/";
  return url;
}
